drivers/timer/nrf_rtc_timer.c

```diff
@@ -550,24 +550,33 @@ int _sys_clock_driver_init(struct device *device)
 
 u32_t _timer_cycle_get_32(void)
 {
-	u32_t elapsed_cycles;
 	u32_t sys_clock_tick_count;
-	u32_t rtc_prev;
-	u32_t rtc_now;
+	u32_t rtc_count;
+	u32_t ticked_cycles;
+	u32_t elapsed_cycles;
 
-	rtc_now = RTC_COUNTER;
-	/* Discard value of RTC_COUNTER read at LFCLK transition */
-	do {
-		sys_clock_tick_count = _sys_clock_tick_count;
-		elapsed_cycles = (rtc_now - (sys_clock_tick_count *
-					     sys_clock_hw_cycles_per_tick)) &
-				 RTC_MASK;
-		rtc_prev = rtc_now;
-		rtc_now = RTC_COUNTER;
-	} while (rtc_now != rtc_prev);
+	/* Get number of announced ticks. */
+	sys_clock_tick_count = _sys_clock_tick_count;
+
+	/* Make sure that compiler will not reverse access to RTC and
+	 * _sys_clock_tick_count
+	 */
+	compiler_barrier();
+
+	/* Get current value of RTC counter. As this function could be
+	 * preempted any time, our sys_clock_tick_count value might be
+	 * outdated at this point. However the value of rtc_count will
+	 * include missed ticks.
+	 */
+	rtc_count = RTC_COUNTER;
```

Comment:
I think that it is compensated by rtc_count. Please note that the `_sys_clock_tick_count` is read before the `RTC_COUNTER` register (this order is enforced by compiler barrier). As `RTC_COUNTER` is _not_ reset on tick, it will include the missed time.

Please look at the following example:
````
RTC_COUNTER value:                        100           200           300
tick:                                                       1               2               3
_sys_clock_tick_count read:                             x
RTC_COUNTER read:                                                       x
````

In this example we we read `_sys_clock_tick_count` somewhere between tick 1 and 2. The read value was 1. Then we got an interrupt, which delayed `RTC_COUNTER` read (which returned 250, as RTC is counting all the time). The final calculation uses cached value of `_sys_clock_tick_count` (which is stale in this example) and current value of `RTC_COUNTER`:

ticked_cycles = cached _sys_clock_tick_count * RTC_COUNTS_PER_TICK = 1 * 100 = 100
elapsed_cycles = RTC_COUNT - ticked_cycles = 250 - 100 = 150
(please note that elapsed_cycles is larger than one tick).

total_time = 100 + 150 = 250  (correct)